mudlib/u/t/tim/makemanpg.6
.\"make a nroff ready man page for a command or function
.TH makemanpg 6 "Oct 7 1995" Ultralib "Wizard Commands"

.SH NAME
makemanpg - make a nroff ready man page for a command or function

.SH SYNOPSIS
makemanpg <section #>

.SH DESCRIPTION
makemanpg(6) is used to produce a nroff ready man page, which can then be
used with Unix's man command, or as nroff source to produce a text file
formatted in the traditional 'man' style.
.PP
makemanpg(6) should be invoked giving the section number as its only argument.
A list of sections is provided at the end of this page.
.PP
When makemanpg(6) is invoked, it prompts the user to enter a command name.
This should be the name of the command/function being documented.  Next,
the user will be prompted to enter a short description for the command/function.
The short description must be entered on one line, and should usually be
less than one line long.  After the short description is entered, the user
will be prompted to enter the SYNOPSIS field.  This field should show how
to the command/function may be invoked, listing all possible forms.  This
field may contain several lines, and a ** should be entered on a line by
itself to denote the end of the section.  Next, the user will be prompted
to enter the DESCRIPTION field.  This field is the 'body'
of the man page, and should describe how to use the command/function.  This
field is unlimited in length, and should be entered in the same manner as
the SYNOPSIS field; enter a ** on a line by
itself to denote the end of the DESCRIPTION section.  Next, the user will be
prompted to enter the SEE ALSO section.  This section should provide 
references to related commands/functions and should be entered in the same
manner as the DESCRIPTION and SYNOPSIS fields, ending with a ** on a line
by itself.
.SH NROFF COMMANDS
When entering any of the fields, raw nroff commands may be included to change
the format of the man page.  Most commands are preceded with a period (.) and
a few are listed in the following section.
.NF
     sp n   insert n spacing lines
     ce n   center next n lines
     na     no alignment of right margin
     PP     begin new paragraph
     SH t   section heading follows; t is the heading
     NF     don't format the next block (?)

.SH OUTPUT
The output file is created in the player's current working directory. (this
is subject to a valid_write(4) so it may not be created)

.SH MAN SECTIONS
.NF
  Section  Topic
  -------  -----
     1     Not Used
     2     Simulated External Functions
     3     LPC Library Functions
     4     Driver Applies
     5     Player Commands
     6     Wizard Commands
     7     Admin Commands

.SH SEE ALSO
man(5), valid_write(4)


mudlib/ChangeLog
Thu Jul 18 23:43:38 1996  MuD Admin Acct  <mud@deathstar.galaxy.net>

	* new commands:
	 . du:  Display disk usage statistics (in Kb)
	 . aliasch: create local aliases for intermud chat channels
	



mudlib/adm/cmd/wiz/makemanpg.c
/*  -*- LPC -*-  */
#include <command.h>
#include <mudlib.h>
#include <time.h>

inherit INHERIT_DIR "/time";

#define USAGE \
"usage: makemanpg <section #>"\
"  sections are:\n"\
"    1: nothing\n"\
"    2: Simulated External Functions\n"\
"    3: LPC Library Functions\n"\
"    4: Driver Applies\n"\
"    5: Player Commands\n"\
"    6: Wizard Commands\n"\
"    7: Admin  Commands"

string *sections = ({
  "",                               // man1
  "Simulated External Functions",   // man2
  "LPC Library Functions",          // man3
  "Driver Applies",                 // man4
  "Player Commands",                // man5
  "Wizard Commands",                // man6
  "Admin Commands",                 // man7
});

void
write_manpg(mapping cmdinfo)
{
  string date = cmdinfo["date"],
         name = cmdinfo["name"],
         short_desc = cmdinfo["short_desc"],
         *synopsis = cmdinfo["synopsis"],
         *description = cmdinfo["description"],
         *see_also = cmdinfo["see_also"];
  int section = cmdinfo["section"];
  string line;
  string outfile;

  outfile = RESOLVE_PATH(name +"."+ section + ".nroff");

  write_file(outfile, ".\\\""+ short_desc +"\n", 1);
  write_file(outfile, ".TH "+ name +" "+ section + " \"" + date +
	     "\" Ultralib \""+ sections[section-1] +"\"\n\n");
  write_file(outfile, ".SH NAME\n");
  write_file(outfile, short_desc + "\n\n");
  write_file(outfile, ".SH SYNOPSIS\n");
  foreach(line in synopsis)
    write_file(outfile, line +"\n");
  write_file(outfile, "\n");
  write_file(outfile, ".SH DESCRIPTION\n");
  foreach(line in description)
    write_file(outfile, line +"\n");
  write_file(outfile, "\n");
  write_file(outfile, ".SH SEE ALSO\n");
  foreach(line in see_also)
    write_file(outfile, line +"\n");
  
  printf("\n Output file %s has been created.\n", outfile);
}
  
int
main(string arg)
{
  mapping cmdinfo = ([]);
  string date = get_date(time(), DATESTR_DATE) +" "+
                get_date(time(), DATESTR_YEAR);

  cmdinfo += ([ "date" : date ]);
  cmdinfo += ([ "synopsis" : ({}) ]);
  cmdinfo += ([ "description" : ({}) ]);
  cmdinfo += ([ "see_also" : ({}) ]);

  if(!arg)
    return notify_fail(USAGE +"\n");

  cmdinfo["section"] = atoi(arg);
  write("File Name (don't use parens): ");
  input_to("get_cmd_name", 0, cmdinfo);
  return 1;
}

void
get_cmd_name(string input, mapping cmdinfo)
{
  cmdinfo["name"] = input;
  write("Enter NAME line--short descrip--not usage (less than 1 line):\n");
  input_to("get_short_desc", 0, cmdinfo);
}

void
get_short_desc(string input, mapping cmdinfo)
{
  cmdinfo["short_desc"] = input;
  write("Enter SYNOPSIS (usage) (usally 1 line); end with ** on a line by itself.\n");
  input_to("get_synopsis", 0, cmdinfo);
}

void
get_synopsis(string input, mapping cmdinfo)
{
  if(input != "**") {
    cmdinfo["synopsis"] += ({ input });
    input_to("get_synopsis", 0, cmdinfo);
    return;
  }
  write("Enter DESCRIPTION of command; end with ** on a line by itself\n");
  input_to("get_description", 0, cmdinfo);
}

void
get_description(string input, mapping cmdinfo)
{
  if(input != "**") {
    cmdinfo["description"] += ({ input });
    input_to("get_description", 0, cmdinfo);
    return;
  }
  write("Enter SEE ALSO for command; end with ** on a line by itself\n");
  input_to("get_see_also", 0, cmdinfo);
}

void
get_see_also(string input, mapping cmdinfo)
{
  if(input != "**") {
    cmdinfo["see_also"] += ({ input });
    input_to("get_see_also", 0, cmdinfo);
    return;
  }
  write_manpg(cmdinfo);
}

string
help()
{
  return "MANPAGE:cat5/makemanpg.6";
}
